packages/wrangler/src/config/validation.ts

```diff
@@ -1142,6 +1152,68 @@ const validateRule: ValidatorFn = (diagnostics, field, value) => {
   return isValid;
 };
 
+const validateDefines =
+  (envName: string): ValidatorFn =>
+  (diagnostics, field, value, config) => {
+    let isValid = true;
+    const fieldPath =
+      config === undefined ? `${field}` : `env.${envName}.${field}`;
+
+    if (typeof value === "object" && value !== null) {
+      for (const varName in value) {
+        // some casting here to appease typescript
+        // even though the value might not match the type
+        if (typeof (value as Record<string, string>)[varName] !== "string") {
+          diagnostics.errors.push(
+            `The field "${fieldPath}.${varName}" should be a string but got ${JSON.stringify(
+              (value as Record<string, string>)[varName]
+            )}.`
+          );
+          isValid = false;
+        }
+      }
+    } else {
+      if (value !== undefined) {
+        diagnostics.errors.push(
+          `The field "${fieldPath}" should be an object but got ${JSON.stringify(
+            value
+          )}.\n`
+        );
+        isValid = false;
+      }
+    }
+
+    const configDefines = Object.keys(config?.define ?? {});
+
+    // If there are no top level vars then there is nothing to do here.
+    if (configDefines.length > 0) {
+      if (typeof value === "object" && value !== null) {
+        const configEnvDefines = config === undefined ? [] : Object.keys(value);
+
+        for (const varName of configDefines) {
+          if (!(varName in value)) {
+            diagnostics.warnings.push(
+              `"define.${varName}" exists at the top level, but not on "${fieldPath}".\n` +
+                `This is not what you probably want, since "define" configuration is not inherited by environments.\n` +
+                `Please add "define.${varName}" to "env.${envName}".`
+            );
+          }
+        }
+        for (const varName of configEnvDefines) {
+          if (!configDefines.includes(varName)) {
+            diagnostics.warnings.push(
+              `"${varName}" exists on "env.${envName}", but not on the top level.\n` +
+                `This is not what you probably want, since "define" configuration is not inherited by environments.\n` +
```

Comment:
> since "define" configuration is not inherited by environments

Struggling to grok this part of the error message. To me, inheritance flows top to bottom, so the error would make sense if they defined it at the top level, but not in an environment.

Do we mean:

> This is not what you probably want, since "define" configuration within environments can only override existing top level "define" configuration

?

Reply:
That's a good suggestion. Will change.

Reply:
done